### PR TITLE
Add ifctrstat applet

### PR DESCRIPTION
**Pull request by awilfox**
Based on original specification [ifstat.rst](https://github.com/ifupdown-ng/ifupdown-ng/files/5066291/ifstat.txt), but renamed to avoid conflict with iproute2 tool of the same name.

### CODE REVIEW

File: cmd/ifctrstat_linux.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"rx.packets",
+	"rx.discard",
+	"rx.errors",
+	"tx.octets",
+	"tx.packets",
+	"tx.discard",
+	"tx.errors"
+};
+
+size_t avail_counters_count = sizeof(avail_counters) / sizeof(*avail_counters);
```
Comment (kaniini) [inline cmd/ifctrstat_linux.c:30]: Use `ARRAY_SIZE()` from `multicall.h`.

File: cmd/ifctrstat_linux.c [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/*
+ * cmd/ifctrstat_linux.c
```
Comment (kaniini) [inline cmd/ifctrstat_linux.c:2]: `cmd/ifctrstat-linux.c`.

File: Makefile [hunk 3]
```diff
@@ -50,6 +50,9 @@ MULTICALL = ifupdown
 IFQUERY_SRC = cmd/ifquery.c
 IFQUERY_OBJ = ${IFQUERY_SRC:.c=.o}
 
+IFCTRSTAT_SRC = cmd/ifctrstat.c cmd/ifctrstat_linux.c
```
Comment (kaniini) [inline Makefile:53]: `cmd/ifctrstat-${LAYOUT}.c`.

File: cmd/ifctrstat.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		return EXIT_FAILURE;
+	}
+
+	const char *iface = argv[idx++];
+
+	for (; idx < argc; idx++)
+	{
+		if (!counter_is_valid(argv[idx]))
+		{
+			fprintf(stderr, "%s: counter %s is not valid or not "
```
Comment (kaniini) [inline cmd/ifctrstat.c:76]: wrap to 200 chars, not 72.

**Comment by awilfox** [comment]
I believe all the requested changes have been implemented.  I will rebase against `master` and make `ifctrstat` configurable.

**Comment by awilfox** [comment]
Okay, that's done.  Please re-review.

File: cmd/ifctrstat-linux.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+{
+	FILE *fp;
+	const char *path;
+	char full_path[PATH_MAX];
+	char buffer[1024];
+	size_t in_count;
+
+	errno = 0;
+
+	if (strcasecmp(counter, "rx.octets") == 0)
```
Comment (kaniini) [inline cmd/ifctrstat-linux.c:46]: Use a mapping struct and bsearch(3) here, please.

File: cmd/ifctrstat.c [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		}
+
+		print_counter(iface, argv[idx], res);
+	}
+
+	return EXIT_SUCCESS;
+}
+
+static struct if_option local_options[] = {
+	{'L', "list", NULL, "List available counters", false, ifctrstat_list_counters},
```
Comment (kaniini) [inline cmd/ifctrstat.c:145]: Using a unified handler with a switch statement will reduce code bloat.

File: Makefile [hunk 7]
```diff
@@ -62,6 +62,13 @@ MULTICALL_${CONFIG_IFQUERY}_OBJ += ${IFQUERY_SRC:.c=.o}
 CMDS_${CONFIG_IFQUERY} += ifquery
 CPPFLAGS_${CONFIG_IFQUERY} += -DCONFIG_IFQUERY
 
+# enable ifctrstat applet (+16 KB)
```
Comment (kaniini) [inline Makefile:65]: is the +16 KB actually accurate here?

Comment (awilfox) [reply Makefile:65]: Yes.  GCC 8 with default CFLAGS (`-Os -ggdb3`).  `CONFIG_IFCTRSTAT=N` vs `Y` changes the size of `ifupdown`:

```
awilcox on catbookpro [pts/0 Fri 14 11:28] ifupdown-ng: ls -al ifupdown
-rwxr-xr-x 1 awilcox awilcox 214040 Aug 14 11:28 ifupdown
awilcox on catbookpro [pts/0 Fri 14 11:28] ifupdown-ng: make CONFIG_IFCTRSTAT=Y
...
awilcox on catbookpro [pts/0 Fri 14 11:28] ifupdown-ng: ls -al ifupdown        
-rwxr-xr-x 1 awilcox awilcox 230888 Aug 14 11:28 ifupdown
```

File: cmd/ifctrstat-linux.c (hunk 5 above)
Comment (awilfox) [reply cmd/ifctrstat-linux.c:46]: Yeah.  I'm thinking of a slightly different design that should reduce size as well.

**Comment by kaniini** [comment]
I've been measuring the stripped version delta.On Aug 14, 2020 8:46 PM, "A. Wilcox" <notifications@github.com> wrote:
@awilfox commented on this pull request.



In Makefile:
> @@ -62,6 +62,13 @@ MULTICALL_${CONFIG_IFQUERY}_OBJ += ${IFQUERY_SRC:.c=.o}
 CMDS_${CONFIG_IFQUERY} += ifquery
 CPPFLAGS_${CONFIG_IFQUERY} += -DCONFIG_IFQUERY
 
+# enable ifctrstat applet (+16 KB)

Yes.  GCC 8 with default CFLAGS (-Os -ggdb3).  CONFIG_IFCTRSTAT=N vs Y changes the size of ifupdown:
awilcox on catbookpro [pts/0 Fri 14 11:28] ifupdown-ng: ls -al ifupdown
-rwxr-xr-x 1 awilcox awilcox 214040 Aug 14 11:28 ifupdown
awilcox on catbookpro [pts/0 Fri 14 11:28] ifupdown-ng: make CONFIG_IFCTRSTAT=Y
...
awilcox on catbookpro [pts/0 Fri 14 11:28] ifupdown-ng: ls -al ifupdown        
-rwxr-xr-x 1 awilcox awilcox 230888 Aug 14 11:28 ifupdown

File: Makefile (hunk 7 above)
Comment (kaniini) [reply Makefile:65]: I've been measuring the stripped version delta.  Most likely this is +4KB at most.

**Pull request merged**